packages/queued-request-controller/src/QueuedRequestController.ts

```diff
@@ -263,30 +282,61 @@ export class QueuedRequestController extends BaseController<
    */
   async #processNextBatch() {
     const firstRequest = this.#requestQueue.shift() as QueuedRequest;
-    this.#originOfCurrentBatch = firstRequest.origin;
-    this.#networkClientIdOfCurrentBatch = firstRequest.networkClientId;
-    const batch = [firstRequest.processRequest];
+    this.#originOfCurrentBatch = firstRequest.request.origin;
+    this.#networkClientIdOfCurrentBatch = firstRequest.request.networkClientId;
+    const batch = [firstRequest];
+
+    let networkSwitchError: unknown;
+    try {
+      // If globally selected network is different from origin selected network,
+      // switch network before processing batch
+      await this.#switchNetworkIfNecessary(
+        firstRequest.request.networkClientId,
+      );
+    } catch (error: unknown) {
+      networkSwitchError = error;
+    }
+
+    // If the first request might switch the network, process the request by
+    // itself. If the request does change the network, clear the queue for the
+    // origin since it any remaining requests are now invalidated
+    if (this.#canRequestSwitchNetworkWithoutApproval(firstRequest.request)) {
+      // This hack prevents the next batch from being processed
+      // after this request returns. This is necessary because
+      // we may need to flush the queue before the next set of requests
+      // are batched and processed, which we cannot do without blocking
+      // the queue from continuing by artificially increasing the processing
+      // request count
+      this.#processingRequestCount += 1;
+      firstRequest.processRequest(networkSwitchError);
```

Comment:
What is the effect of passing `networkSwitchError` to `processRequest` here and do we have a test for this?

Reply:
Note that this was the previous behavior as well (see below where `processRequest` was already called).

The relevant test is in the `when the network switch for a batch fails` block, the test is called `throws error`. Though that's for the general case. Maybe worth adding a test case to ensure it works correctly here as well (when processing a request that can switch chains without approval).